Implement multiple examples loading and generation in the Interactive Server.

StarKhan6368 opened the pull request:
**What**: Implement multiple examples loading and generation in the Interactive Server.

Load multiple examples and generate more examples in interactive examples UI.

**Why**: 
- At present, we are limited to loading only a single existing example, regardless of the number available. As a result, the user can view just that one example and is restricted to generating only one additional example if no examples are present.
- This PR allows for the loading of multiple existing examples and provides the user with the capability to generate additional examples if necessary.

<!-- How were these changes implemented? -->
- Adding `Generate More` button when there are already existing examples
- Show each example as a separate row under each operation

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation
- [x] Tests
- [ ] Sonar Quality Gate

<!-- feel free to add additional comments -->

StarKhan6368 committed:
Modify the logic when creating test from example.

- Fixes - bug with exactValue validation in interactive example tests.
- Use the fileSource attribute in rows to match example file to scenario
- Add tests to ensure canonical path is used for matching.
- Update existing test for Feature.createContractTestFromExampleFile.

StarKhan6368 committed:
Feature - multiple examples in Interactive Server.

- Use AtomicInteger for unique example file names.
- Include TestResult report in case of test failure with testLogs.
- Ability to Load and display multiple existing examples.
- Ability to Generate additional examples when necessary.
- Update ExamplesCommandTest accordingly.

StarKhan6368 committed:
Major update frontend to handle multiple examples.

- Update CSS and JS files.
- Remove chevronDown and dropdown functionality from drill-downs.
- Add and implement Generate more buttons for multiple generations
- Functions to dynamically update the examples table and recalculate
  the row-spans when adding new examples.
- Use dynamic serial numbering no reliance on thymeleaf.
- Modifications to how data is stored and accessed in JS.
- Clean up TestInteractionsLog.combineLog for consistency.

StarKhan6368 committed:
Fix rows creation differing only by content-type.

- Group also by content-type when grouping endpoints.
- Update updateSpans function in JS to also consider content-type.
- Other CSS fixes.

harikrishnan83 committed:
Merge branch 'main' into multiple_examples

* main:
  Loading config file from classpath Also in case of env property or classpath checking if the file exists before returning

harikrishnan83 commented on core/src/main/kotlin/io/specmatic/core/examples/server/ExamplesView.kt:
```diff
@@ -14,56 +16,81 @@ class ExamplesView {
             val examples = examplesDir.getExamplesFromDir()
-            return feature.scenarios.map {
-                val example = getExistingExampleFile(it, examples)
+            val scenarioExPairList = getScenarioExamplesPairs(feature.scenarios, examples)
```
Please expand `Ex` to `Example`, will be easier to read

harikrishnan83 commented on core/src/main/kotlin/io/specmatic/core/examples/server/ExamplesView.kt:
```diff
@@ -14,56 +16,81 @@ class ExamplesView {
-                val example = getExistingExampleFile(it, examples)
+            val scenarioExPairList = getScenarioExamplesPairs(feature.scenarios, examples)
+
+            return scenarioExPairList.map { (scenario, ex) ->
```
same as above, expand `ex` to `example`

harikrishnan83 commented on core/src/main/kotlin/io/specmatic/core/examples/server/ExamplesView.kt:
```diff
@@ -14,56 +16,81 @@ class ExamplesView {
+                    method = scenario.method,
+                    responseStatus = scenario.httpResponsePattern.status,
+                    contentType = scenario.httpRequestPattern.headersPattern.contentType,
+                    exampleFile = ex?.first,
```
Can we further explode `ex` to `exampleFile` and `exampleMismatchReason`? so that we avoid using `first` and `second`

harikrishnan83 commented on core/src/main/kotlin/io/specmatic/core/examples/server/ExamplesView.kt:
```diff
@@ -14,56 +16,81 @@ class ExamplesView {
+                    exampleFile = ex?.first,
+                    exampleMismatchReason = ex?.second
                 )
-            }.filterEndpoints().sortEndpoints()
```
any reason we stopped sorting endpoints? was this intentional?

harikrishnan83 committed:
Renamed variables for readability

StarKhan6368 committed:
Fix - invalid attribute typo to `failed`.

StarKhan6368 committed:
Revert - Modifications to createTestFromExample

- This change will be incorporated into a separate PR.

StarKhan6368 committed:
Rename example file name counter and reset method.

- Rename `atomicCounter` to `exampleFileNamePostFixCounter` to better
  describe the functionality.

harikrishnan83 reviewed (approved):
Looks good to me
